### PR TITLE
Fix handling maps in matchers

### PR DESCRIPTION
**Pull request by wpanas**

### CODE REVIEW

File: hermes-common/src/test/groovy/pl/allegro/tech/hermes/domain/filtering/AvroPathMessageFilterSpec.groovy [hunk 1]
```diff
@@ -47,22 +50,23 @@ class AvroPathMessageFilterSpec extends Specification {
...
-        ".batter.id"              | "1003"      | true
-        ".batter.id"              | "1004"      | false
-        ".topping.type"           | "^Map.*"    | true
-        ".topping.description"    | ".*syrup.*" | true
-        ".topping.description.a"  | ".*"        | false
-        ".batter.ingredients"     | "null"      | true
-        ".batter.ingredients"     | ".*"        | true
-        ".batter.ingredients"     | "sugar"     | false
-        ".topping.ingredients"    | ".*syrup.*" | true
+        path                     | matcher     | result
```
Comment (lukasz20p) [inline hermes-common/src/test/groovy/pl/allegro/tech/hermes/domain/filtering/AvroPathMessageFilterSpec.groovy:53]: Formatowanie się rozjechało.

File: hermes-common/src/test/groovy/pl/allegro/tech/hermes/domain/filtering/AvroPathMessageFilterSpec.groovy [hunk 2]
```diff
@@ -166,10 +170,10 @@ class AvroPathMessageFilterSpec extends Specification {
         when:
         new AvroPathSubscriptionMessageFilterCompiler().compile(new MessageFilterSpecification([path: ".id", matcher: "0001"]))
```
Comment (lukasz20p) [inline hermes-common/src/test/groovy/pl/allegro/tech/hermes/domain/filtering/AvroPathMessageFilterSpec.groovy:171]: Tutaj także formatownie.

File: hermes-common/src/main/java/pl/allegro/tech/hermes/domain/filtering/avro/AvroPathPredicate.java [hunk 3]
```diff
@@ -70,29 +73,37 @@ public boolean test(final FilterableMessage message) {
...
-                if (idx.equals(WILDCARD_IDX)) {
-                    return selectMultipleArrayItems(iter, currentArray);
                 } else {
-                    current = selectSingleArrayItem(Integer.valueOf(idx), currentArray);
+                    current = currentRecord.get(selector);
                 }
-
-            } else {
+            } else if (current instanceof HashMap) {
+                Map<Utf8, Object> currentRecord = (HashMap<Utf8, Object>) current;
```
Comment (lukasz20p) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/domain/filtering/avro/AvroPathPredicate.java:102]: Skąd pewność, że zawsze tutaj będzie Utf8 oraz Object ?

Comment (wpanas) [reply hermes-common/src/main/java/pl/allegro/tech/hermes/domain/filtering/avro/AvroPathPredicate.java:102]: `Object` zawsze będzie, bo wszystko dziedziczy po `Object`, a kluczem w `Map` zakładam, że będzie `Utf8`, bo tak jest deserializowany string w przypadku JSON'a. W naszym przypadku robimy z JSONa AVRO, więc nie zrobisz bardziej złożonego klucza głównego niż string. A jak robisz liczbę to zrób po prostu array.

Comment (pitagoras3) [reply hermes-common/src/main/java/pl/allegro/tech/hermes/domain/filtering/avro/AvroPathPredicate.java:102]: @lukasz20p I think that's because of [Avro standard](https://avro.apache.org/docs/current/spec.html#Maps):
```
Maps
Maps use the type name "map" and support one attribute:

values: the schema of the map's values.
Map keys are assumed to be strings.

For example, a map from string to long is declared with:

{
  "type": "map",
  "values" : "long",
  "default": {}
}
```

For example, metadata is processed as `Map<Utf8, Utf8> metadata`

File: hermes-common/src/main/java/pl/allegro/tech/hermes/domain/filtering/avro/AvroPathPredicate.java [hunk 4]
```diff
@@ -70,29 +73,37 @@ public boolean test(final FilterableMessage message) {
...
+                Utf8 selector = new Utf8(iter.next());
                 current = currentRecord.get(selector);
+            } else {
+                break;
```
Comment (pitagoras3) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/domain/filtering/avro/AvroPathPredicate.java:106]: Note - this break is substitute for `while (iter.hasNext() && (current instanceof GenericRecord) || current instanceof HashMap))`. I racked my brain for a bit to get into that. What do you think about getting rid of this `else { break; }` and to use this `iter.hasNext() && (current instanceof GenericRecord) || current instanceof HashMap)` boolean expression (and extracting this boolean expression to separate function)?

File: hermes-common/src/test/groovy/pl/allegro/tech/hermes/domain/filtering/AvroPathMessageFilterSpec.groovy [hunk 5]
```diff
@@ -20,6 +20,9 @@ class AvroPathMessageFilterSpec extends Specification {
                 "id": "0001",
                 "type": "donut",
                 "name": "Cake",
+                "flag": {
```
Comment (pitagoras3) [inline hermes-common/src/test/groovy/pl/allegro/tech/hermes/domain/filtering/AvroPathMessageFilterSpec.groovy:23]: Can we also make a test for empty map and for map containing more than 1 element?

Comment (lukasz20p) [reply hermes-common/src/test/groovy/pl/allegro/tech/hermes/domain/filtering/AvroPathMessageFilterSpec.groovy]: Ofc. Done

File: hermes-common/src/main/java/pl/allegro/tech/hermes/domain/filtering/avro/AvroPathPredicate.java (hunk 4 above)
Comment (lukasz20p) [reply hermes-common/src/main/java/pl/allegro/tech/hermes/domain/filtering/avro/AvroPathPredicate.java]: I add new method for decide if analyzed record type is supported

**Review by pitagoras3 — APPROVED** [review]
